boot/seal.go

```diff
@@ -211,24 +275,87 @@ func resealKeyToModeenv(rootdir string, model *asserts.Model, modeenv *Modeenv,
 		bootloader.RoleRunMode:  bl.Name(),
 	}
 
+	authKeyFile := filepath.Join(dirs.SnapFDEDirUnder(rootdir), "tpm-policy-auth-key")
+	if err := resealRunObjectKeys(pbc, authKeyFile, roleToBlName); err != nil {
+		return err
+	}
+	logger.Debugf("resealing (%d) succeeded", nextCount)
+
+	// reseal the fallback object
+	rpbc := toPredictableBootChains(recoveryBootChains)
+
+	var nextFallbackCount int
+	needed, nextFallbackCount, err = isFallbackResealNeeded(rpbc, rootdir, expectReseal)
+	if err != nil {
+		return err
+	}
+	if needed {
+		rpbcJSON, _ := json.Marshal(rpbc)
+		logger.Debugf("resealing (%d) to recovery boot chains: %s", nextCount, rpbcJSON)
+
+		roleToBlName := map[bootloader.Role]string{
+			bootloader.RoleRecovery: rbl.Name(),
+			bootloader.RoleRunMode:  bl.Name(),
+		}
+
+		if err := resealRunObjectKeys(pbc, authKeyFile, roleToBlName); err != nil {
+			return err
+		}
+		logger.Debugf("fallback resealing (%d) succeeded", nextFallbackCount)
+	}
+
+	bootChainsPath := bootChainsFileUnder(rootdir)
+	if err := writeBootChains(pbc, rpbc, bootChainsPath, nextCount, nextFallbackCount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resealRunObjectKeys(pbc predictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
 	// get model parameters from bootchains
 	modelParams, err := sealKeyModelParams(pbc, roleToBlName)
 	if err != nil {
 		return fmt.Errorf("cannot prepare for key resealing: %v", err)
 	}
+
+	// list all the key files to reseal
+	keyFiles := []string{
+		filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
+	}
+
 	resealKeyParams := &secboot.ResealKeyParams{
 		ModelParams:          modelParams,
-		KeyFile:              filepath.Join(InitramfsEncryptionKeyDir, "ubuntu-data.sealed-key"),
-		TPMPolicyAuthKeyFile: filepath.Join(dirs.SnapFDEDirUnder(rootdir), "tpm-policy-auth-key"),
+		KeyFiles:             keyFiles,
+		TPMPolicyAuthKeyFile: authKeyFile,
 	}
 	if err := secbootResealKey(resealKeyParams); err != nil {
 		return fmt.Errorf("cannot reseal the encryption key: %v", err)
 	}
-	logger.Debugf("resealing (%d) succeeded", nextCount)
 
-	bootChainsPath := bootChainsFileUnder(rootdir)
-	if err := writeBootChains(pbc, bootChainsPath, nextCount); err != nil {
-		return err
+	return nil
+}
+
+func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, roleToBlName map[bootloader.Role]string) error {
```

Comment:
This does not appear to be called anywhere?

Reply:
There was a bug in the code where the run object reseal was being called instead. Fixing.